src/discovery/Discovery.ts

```diff
@@ -175,299 +197,291 @@ class Discovery {
+                linkedVertexNodeId,
+              );
+            } catch (e) {
+              if (
+                e instanceof nodesErrors.ErrorNodeConnectionDestroyed ||
+                e instanceof nodesErrors.ErrorNodeConnectionTimeout ||
+                e instanceof networkErrors.ErrorConnectionNotRunning
+              ) {
                 if (!this.visitedVertices.has(linkedVertexGK)) {
-                  await this.pushKeyToDiscoveryQueue(linkedVertexGK);
+                  await this.startDiscoveryForVertex(linkedVertexGK);
                 }
+                this.logger.error(
+                  `Failed to discover ${data.node} - ${e.toString()}`,
+                );
+                continue;
+              } else {
+                throw e;
               }
             }
-            break;
-          default:
-            never();
+            // With this verified chain, we can link
+            const linkedVertexNodeInfo: NodeInfo = {
+              id: nodesUtils.encodeNodeId(linkedVertexNodeId),
+              chain: linkedVertexChainData,
+            };
+            await this.gestaltGraph.linkNodeAndIdentity(
+              linkedVertexNodeInfo,
+              vertexIdentityInfo,
+            );
+            // Add this vertex to the queue if it is not present
+            if (!this.visitedVertices.has(linkedVertexGK)) {
+              await this.startDiscoveryForVertex(linkedVertexGK);
+            }
+          }
         }
-        this.visitedVertices.add(vertex);
-        await this.removeKeyFromDiscoveryQueue(vertexId);
-      }
+        break;
+      default:
+        never();
     }
+    this.visitedVertices.add(vertex);
   }
 
   /**
-   * Will resolve once the queue has drained
+   * Will resolve once all discovery tasks have finished
    */
   public async waitForDrained(): Promise<void> {
-    await this.queueDrained.p;
+    let promises: Array<Promise<any>>;
+    do {
+      promises = [];
+      for await (const task of this.taskManager.getTasks('asc', false, [
+        this.constructor.name,
+        this.discoverVertexHandlerId,
+      ])) {
+        promises.push(task.promise());
+      }
+      await Promise.all(promises);
+    } while (promises.length > 0);
   }
 
   /**
    * Simple check for whether the Discovery Queue is empty.
    */
   protected async queueIsEmpty(): Promise<boolean> {
-    let nextDiscoveryQueueId: DiscoveryQueueId | undefined;
-    const keyIterator = this.db.iterator(this.discoveryQueueDbPath, {
-      limit: 1,
-      values: false,
-    });
-    for await (const [keyPath] of keyIterator) {
-      const key = keyPath[0] as Buffer;
-      nextDiscoveryQueueId = IdInternal.fromBuffer<DiscoveryQueueId>(key);
+    for await (const _ of this.taskManager.getTasks('asc', true, [
+      this.constructor.name,
+      this.discoverVertexHandlerId,
+    ])) {
+      return false;
     }
-    return nextDiscoveryQueueId == null;
+    return true;
   }
```

Comment:
Not used anywhere, might remove it?

Reply:
I'll leave it, It may be useful.

Reply:
Comment seems incorrect. There is no more discovery queue. And what we are checking here is the presence of discovery tasks. Perhaps this method should be renamed? And comment rewritten.

Reply:
`hasPendingDiscovery`, `discoveryIsActive`? not sure what to call it.

Reply:
Since we use `waitForDrained` maybe call it `isDrained` to complement it?

Reply:
This has been updated.